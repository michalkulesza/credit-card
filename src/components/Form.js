import React from "react";
import "./Form.scss";

const Form = ({
  formatCard,
  ccNum,
  setCcNum,
  maxLength,
  name,
  setName,
  month,
  setMonth,
  year,
  setYear,
  setCvvActive,
  cvv,
  setCvv,
  setCursor
}) => {
  let years = [];

  const handleOnChangeCardNumber = e => {
    let cursor = e.target.selectionStart;
    let lastVal = e.target.value;
    let formattedNum = formatCard(e.target.value);

    setCcNum(formattedNum);

    if (cursor === lastVal.length) {
      cursor = formattedNum.length;
      if (
        e.target.getAttribute("data-lastvalue") &&
        e.target.getAttribute("data-lastvalue").charAt(cursor - 1) === " "
      ) {
        cursor--;
      }
    }

    if (lastVal !== formattedNum) {
      if (lastVal.charAt(cursor) === " " && formattedNum.charAt(cursor - 1) === " ") {
        cursor++;
      }
    }

    e.target.selectionStart = cursor;
    e.target.selectionEnd = cursor;
    e.target.setAttribute("data-lastvalue", formattedNum);
    setCursor(cursor);
  };

  const handleOnChangeName = e => {
    setName(e.target.value.toUpperCase().replace(/[^a-zA-Z ]/g, ""));
  };

  for (let i = 0; i < 10; i++) {
    let year = new Date().getFullYear();
    years.push(<option key={Math.random() * i}>{year + i}</option>);
  }

  return (
    <div className="form-wrapper">
      <div className="form-row row-1">
        <label htmlFor="cardNumber">Card Number</label>
        <input
          type="text"
          id="cardNumber"
          onChange={handleOnChangeCardNumber}
          value={ccNum}
          maxLength={maxLength}
          className="card-number-input"
        />
      </div>
      <div className="form-row">
        <label htmlFor="cardHolder">Card Holders Name</label>
        <input
          type="text"
          id="cardHolder"
          className="card-holder-input"
          onChange={handleOnChangeName}
          value={name}
          maxLength="31"
        />
      </div>
      <div className="form-row row-3">
        <div className="card-expires">
          <label htmlFor="cardExpires">Expiration Date</label>
          <div className="card-expires-selection">
            <select
              value={month !== 0 ? month : "Month"}
              onChange={e => setMonth(e.target.value)}
              id="cardMonth"
              className="card-expires-month"
            >
              <option disabled="disabled">Month</option>
              <option>01</option>
              <option>02</option>
              <option>03</option>
              <option>04</option>
              <option>05</option>
              <option>06</option>
              <option>07</option>
              <option>08</option>
              <option>09</option>
              <option>10</option>
              <option>11</option>
              <option>12</option>
            </select>
            <select
              value={year !== 0 ? year : "Year"}
              onChange={e => setYear(e.target.value)}
              id="cardYear"
              className="card-expires-year"
            >
              <option disabled="disabled">Year</option>
              {years}
            </select>
          </div>
        </div>
        <div className="card-cvv">
          <label htmlFor="cardCvv">CVV</label>
          <input
            onFocus={() => setCvvActive(true)}
            onBlur={() => setCvvActive(false)}
            type="text"
            id="cardCvv"
            className="card-holder-cvv"
            maxLength="4"
            onChange={e => setCvv(e.target.value.toString().replace(/\D/g, ""))}
            value={cvv}
          />
        </div>
      </div>
      <div className="form-row"></div>
    </div>
  );
};

export default Form;
