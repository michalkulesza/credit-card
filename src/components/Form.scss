.form-wrapper {
  height: 550px;
  width: 570px;
  background-color: white;
  border-radius: 10px;
  padding: 35px;
  padding-top: 180px;
  box-shadow: 0 0 10px 10px rgba($color: #000000, $alpha: 0.025);
}

.form-row {
  height: 25%;
  width: 100%;
  display: flex;
  flex-direction: column;

  label {
    height: auto;
    font-size: 0.8em;
    font-weight: 400;
    margin-bottom: 3px;
    color: #1a3b5d;
  }

  input {
    height: auto;
    font-size: 1em;
    padding: 15px 20px;
    border-radius: 5px;
    border: 1px solid #c9d1da;
    color: #1a3b5d;
  }

  &.row-3 {
    flex-direction: row;
    justify-content: space-between;

    .card-expires {
      display: flex;
      flex-direction: column;
      height: auto;

      .card-expires-selection {
        height: auto;
      }
    }

    .card-expires-month {
      margin-right: 10px;
    }

    .card-expires-month,
    .card-expires-year {
      padding: 15px 40px 15px 15px;
      font-size: 1em;
      border-radius: 5px;
      border: 1px solid #c9d1da;
      color: #1a3b5d;
    }

    .card-cvv {
      display: flex;
      flex-direction: column;
    }
    .card-holder-cvv {
      width: 79px;
      text-align: center;
    }
  }
}
