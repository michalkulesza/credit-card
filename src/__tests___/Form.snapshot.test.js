import React from "react";
import Form from "../components/Form";
import renderer from "react-test-renderer";

describe("Form component renders correctly with default props", () => {
  it("renders correctly", () => {
    const rendered = renderer.create(<Form />);
    expect(rendered.toJSON()).toMatchSnapshot();
  });
});
