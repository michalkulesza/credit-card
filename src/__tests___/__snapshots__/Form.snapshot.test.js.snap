// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component renders correctly with default props renders correctly 1`] = `
<div
  className="form-wrapper"
>
  <div
    className="form-row row-1"
  >
    <label
      htmlFor="cardNumber"
    >
      Card Number
    </label>
    <input
      className="card-number-input"
      id="cardNumber"
      maxLength={19}
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="form-row"
  >
    <label
      htmlFor="cardHolder"
    >
      Card Holders Name
    </label>
    <input
      className="card-holder-input"
      id="cardHolder"
      maxLength="31"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="form-row row-3"
  >
    <div
      className="card-expires"
    >
      <label
        htmlFor="cardExpires"
      >
        Expiration Date
      </label>
      <div
        className="card-expires-selection"
      >
        <select
          className="card-expires-month"
          id="cardMonth"
          onChange={[Function]}
        >
          <option
            disabled="disabled"
          >
            Month
          </option>
          <option>
            01
          </option>
          <option>
            02
          </option>
          <option>
            03
          </option>
          <option>
            04
          </option>
          <option>
            05
          </option>
          <option>
            06
          </option>
          <option>
            07
          </option>
          <option>
            08
          </option>
          <option>
            09
          </option>
          <option>
            10
          </option>
          <option>
            11
          </option>
          <option>
            12
          </option>
        </select>
        <select
          className="card-expires-year"
          id="cardYear"
          onChange={[Function]}
        >
          <option
            disabled="disabled"
          >
            Year
          </option>
          <option>
            2020
          </option>
          <option>
            2021
          </option>
          <option>
            2022
          </option>
          <option>
            2023
          </option>
          <option>
            2024
          </option>
          <option>
            2025
          </option>
          <option>
            2026
          </option>
          <option>
            2027
          </option>
          <option>
            2028
          </option>
          <option>
            2029
          </option>
        </select>
      </div>
    </div>
    <div
      className="card-cvv"
    >
      <label
        htmlFor="cardCvv"
      >
        CVV
      </label>
      <input
        className="card-holder-cvv"
        id="cardCvv"
        maxLength="3"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="text"
      />
    </div>
  </div>
  <div
    className="form-row"
  />
</div>
`;
